File: 102. Binary Tree Level Order Traversal.py
# https://leetcode.com/problems/binary-tree-level-order-traversal/

# Given the root of a binary tree, return the level order traversal of its nodes'
# values. (i.e., from left to right, level by level).

################################################################################

# bfs

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque
class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if root == None: return []
        
        ans = []
        queue = deque([root])
                
        while queue:
            curr_level = []
            curr_level_size = len(queue)
            for _ in range(curr_level_size):
                # append node to curr_level
                node = queue.popleft()
                curr_level.append(node.val)
                # append child to queue
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            
            ans.append(curr_level)
        return ans

File: 102. Binary Tree Level Order Traversal.cpp
// https://leetcode.com/problems/binary-tree-level-order-traversal/

// Given the root of a binary tree, return the level order traversal of its nodes'
// values. (i.e., from left to right, level by level).

////////////////////////////////////////////////////////////////////////////////

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (!root) return {};
        
        vector<vector<int>> ans;
        queue<TreeNode*> q; q.push(root);
        vector<int> thisLayer;
        while (!q.empty()) {
            int nThisLayer = q.size();
            while (nThisLayer--) {
                TreeNode* node = q.front(); q.pop();
                thisLayer.push_back(node->val);
                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }
            ans.push_back(thisLayer);
            thisLayer.clear();
        }
        return ans;
    }
};
